src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/PrisonerInEstablishmentService.kt

```diff
@@ -0,0 +1,197 @@
+import java.util.concurrent.TimeUnit
+
+@Service
+class PrisonerInEstablishmentService(
+  private val elasticsearchClient: SearchClient,
+  private val indexStatusService: IndexStatusService,
+  private val gson: Gson,
+  private val telemetryClient: TelemetryClient,
+  private val authenticationHolder: AuthenticationHolder,
+  @Value("\${search.keyword.max-results}") private val maxSearchResults: Int = 200,
+  @Value("\${search.keyword.timeout-seconds}") private val searchTimeoutSeconds: Long = 10L,
+) {
+  companion object {
+    val log: Logger = LoggerFactory.getLogger(this::class.java)
+  }
+
+  fun search(prisonId: String, prisonerSearchRequest: PrisonerInEstablishmentRequest): Page<Prisoner> {
+    log.info("Received keyword request ${gson.toJson(prisonerSearchRequest)}")
+
+    val searchSourceBuilder = createSourceBuilder(prisonId, prisonerSearchRequest)
+    val searchRequest = SearchRequest(arrayOf(getIndex()), searchSourceBuilder)
+
+    // Useful for logging the JSON elastic search query that is executed
+    log.info("search query JSON: {}", searchSourceBuilder.toString())
+
+    val searchResponse = elasticsearchClient.search(searchRequest)
+    return createSearchResponse(prisonerSearchRequest.pagination, searchResponse).also {
+      auditSearch(prisonerSearchRequest, searchResponse.hits.totalHits?.value ?: 0)
+    }
+  }
+
+  private fun createSourceBuilder(
+    prisonId: String,
+    searchRequest: PrisonerInEstablishmentRequest
+  ): SearchSourceBuilder {
+    val pageable = PageRequest.of(searchRequest.pagination.page, searchRequest.pagination.size)
+    return SearchSourceBuilder().apply {
+      timeout(TimeValue(searchTimeoutSeconds, TimeUnit.SECONDS))
+      size(pageable.pageSize.coerceAtMost(maxSearchResults))
+      from(pageable.offset.toInt())
+      trackTotalHits(true)
+      query(buildKeywordQuery(prisonId, searchRequest))
+      sort("lastName.keyword")
+      sort("firstName.keyword")
+      sort("prisonerNumber")
+    }
+  }
+
+  private fun buildKeywordQuery(prisonId: String, searchRequest: PrisonerInEstablishmentRequest): BoolQueryBuilder {
+    val keywordQuery = QueryBuilders.boolQuery()
+
+    // Pattern match terms which might be NomsId, PNC or CRO & uppercase them
+    val sanitisedKeywordRequest = PrisonerInEstablishmentRequest(
+      term = addUppercaseKeywordTokens(searchRequest.term),
+      pagination = searchRequest.pagination,
+    )
+
+    with(sanitisedKeywordRequest) {
+
+      term.takeIf { !it.isNullOrBlank() }?.let {
+        // Will include the prisoner document if any of the words specified match in any of the fields
+        keywordQuery.must().add(
+          generateMatchQuery(it)
+        )
+      }
+
+      keywordQuery.filterWhenPresent("prisonId", prisonId)
+    }
+
+    return keywordQuery
+  }
+
+  private fun generateMatchQuery(
+    term: String,
+  ): QueryBuilder {
+    val fields = listOf(
+      "prisonerNumber",
```

Comment:
list is a bit OTT, not sure sure we need to support for instance CRO for searching within establishment, this will no doubt change